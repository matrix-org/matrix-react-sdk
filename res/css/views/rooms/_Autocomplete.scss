.mx_Autocomplete {
    position: absolute;
    bottom: 0;
    z-index: 1001;
    width: 100%;
    border: 1px solid $primary-hairline-color;
    background: $primary-bg-color;
    border-bottom: none;
    border-radius: 4px 4px 0 0;
    max-height: 50vh;
    overflow: auto;
}

.mx_Autocomplete_ProviderSection {
    border-bottom: 1px solid $primary-hairline-color;
}

/* a "block" completion takes up a whole line */
.mx_Autocomplete_Completion_block {
    height: 34px;
    display: flex;
    padding: 0 12px;
    user-select: none;
    cursor: pointer;
    align-items: center;
    color: $primary-fg-color;
}

.mx_Autocomplete_Completion_block * {
    margin: 0 3px;
}

.mx_Autocomplete_Completion_pill {
    box-sizing: border-box;
    border-radius: 2rem;
    height: $font-34px;
    padding: 0.4rem;
    display: flex;
    user-select: none;
    cursor: pointer;
    align-items: center;
    color: $primary-fg-color;
}

.mx_Autocomplete_Completion_pill > * {
    margin-right: 0.3rem;
}

/* styling for common completion elements */
.mx_Autocomplete_Completion_subtitle {
    font-style: italic;
    flex: 1;
}

.mx_Autocomplete_Completion_description {
    color: gray;
}

.mx_Autocomplete_Completion_container_pill {
    margin: 12px;
    display: flex;
    flex-flow: wrap;
}

.mx_Autocomplete_Completion_container_truncate {
    .mx_Autocomplete_Completion_title,
    .mx_Autocomplete_Completion_subtitle,
    .mx_Autocomplete_Completion_description {
        /* Ellipsis for long names/subtitles/descriptions */
        max-width: 150px;
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
    }
}

.mx_Autocomplete_Completion.selected,
.mx_Autocomplete_Completion:hover {
    background: $selected-color;
    outline: none;
}

.mx_Autocomplete_provider_name {
    margin: 12px;
    color: $primary-fg-color;
    font-weight: 400;
    opacity: 0.4;
}
