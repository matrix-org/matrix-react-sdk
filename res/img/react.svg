<svg xmlns="http://www.w3.org/2000/svg" width="18" height="17" viewBox="0 0 18 17">
    <g fill="none" fill-rule="evenodd">
        <path fill="#2E2F32" fill-rule="nonzero" d="M13.6 14.5a7.875 7.875 0 1 1 2.183-3h-.818a7.125 7.125 0 1 0-2.62 3H13.6z"/>
        <g stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75">
            <path d="M11.5 13h6M14.5 10v6"/>
        </g>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75" d="M5.5 10s1.125 1.5 3 1.5 3-1.5 3-1.5"/>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M6.25 6.25h.008M10.75 6.25h.008"/>
    </g>
</svg>
