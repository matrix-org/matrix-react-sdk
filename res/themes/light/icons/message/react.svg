<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 464c-119.1 0-216-96.9-216-216S128.9 40 248 40s216 96.9 216 216-96.9 216-216 216zm90.2-146.2C315.8 352.6 282.9 368 248 368s-67.8-15.4-90.2-42.2c-5.7-6.8-15.8-7.7-22.5-2-6.8 5.7-7.7 15.7-2 22.5C161.7 380.4 203.6 400 248 400s86.3-19.6 114.8-53.8c5.7-6.8 4.8-16.9-2-22.5-6.8-5.6-16.9-4.7-22.6 2.1zM168 240c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32zm160 0c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32z"/></svg>