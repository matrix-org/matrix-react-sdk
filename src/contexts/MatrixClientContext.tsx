/*
Copyright 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, {
    ComponentClass,
    createContext,
    forwardRef,
    PropsWithoutRef,
    ForwardRefExoticComponent,
    useContext,
    RefAttributes,
} from "react";
import { MatrixClient } from "matrix-js-sdk/src/matrix";

// This context is available to components under LoggedInView,
// the context must not be used by components outside a MatrixClientContext tree.
// This assertion allows us to make the type not nullable.
const MatrixClientContext = createContext<MatrixClient>(null as any);
MatrixClientContext.displayName = "MatrixClientContext";
export default MatrixClientContext;

export interface MatrixClientProps {
    mxClient: MatrixClient;
}

export function useMatrixClientContext(): MatrixClient {
    return useContext(MatrixClientContext);
}

const matrixHOC = <ComposedComponentProps extends {}>(
    ComposedComponent: ComponentClass<ComposedComponentProps>,
): ForwardRefExoticComponent<
    PropsWithoutRef<Omit<ComposedComponentProps, "mxClient">> &
        RefAttributes<InstanceType<ComponentClass<ComposedComponentProps>>>
> => {
    type ComposedComponentInstance = InstanceType<typeof ComposedComponent>;

    // eslint-disable-next-line react-hooks/rules-of-hooks

    const TypedComponent = ComposedComponent;

    return forwardRef<ComposedComponentInstance, Omit<ComposedComponentProps, "mxClient">>((props, ref) => {
        const client = useContext(MatrixClientContext);

        // @ts-ignore
        return <TypedComponent ref={ref} {...props} mxClient={client} />;
    });
};
export const withMatrixClientHOC = matrixHOC;
