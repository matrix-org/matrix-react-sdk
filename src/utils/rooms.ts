/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { MatrixClient } from "matrix-js-sdk/src/matrix";

import { shouldForceDisableEncryption } from "./crypto/shouldForceDisableEncryption";
import { getE2EEWellKnown } from "./WellKnownUtils";

export function privateShouldBeEncrypted(client: MatrixClient): boolean {
    if (shouldForceDisableEncryption(client)) {
        return false;
    }
    const e2eeWellKnown = getE2EEWellKnown(client);
    if (e2eeWellKnown) {
        const defaultDisabled = e2eeWellKnown["default"] === false;
        return !defaultDisabled;
    }
    return true;
}
