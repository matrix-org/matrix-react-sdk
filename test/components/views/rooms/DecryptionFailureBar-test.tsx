/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { act, fireEvent, render, screen, waitFor, RenderResult } from "@testing-library/react";
import "@testing-library/jest-dom";

import MatrixClientContext from "../../../../src/contexts/MatrixClientContext";
import { DecryptionFailureBar } from "../../../../src/components/views/rooms/DecryptionFailureBar";

type MockDevice = { deviceId: string };

const verifiedDevice1: MockDevice = { deviceId: "verified1" };
const verifiedDevice2: MockDevice = { deviceId: "verified2" };
const unverifiedDevice1: MockDevice = { deviceId: "unverified1" };
const unverifiedDevice2: MockDevice = { deviceId: "unverified2" };

const mockEvent1 = {
    event: { event_id: "mockEvent1" },
    getWireContent: () => ({ session_id: "sessionA" }),
};

const mockEvent2 = {
    event: { event_id: "mockEvent2" },
    getWireContent: () => ({ session_id: "sessionB" }),
};

const mockEvent3 = {
    event: { event_id: "mockEvent3" },
    getWireContent: () => ({ session_id: "sessionB" }),
};

const userId = "@user:example.com";

let ourDevice: MockDevice | undefined;
let allDevices: MockDevice[] | undefined;
let keyBackup = false;
let callback = async () => {};

const mockClient = {
    getUserId: () => userId,
    getDeviceId: () => ourDevice?.deviceId,
    getStoredDevicesForUser: () => allDevices,
    isSecretStored: jest.fn(() => Promise.resolve(keyBackup ? { key: "yes" } : null)),
    checkIfOwnDeviceCrossSigned: (deviceId: string) => deviceId.startsWith("verified"),
    downloadKeys: jest.fn(() => {}),
    cancelAndResendEventRoomKeyRequest: jest.fn(() => {}),
    on: (_: any, cb: () => Promise<void>) => {
        callback = cb;
    },
    off: () => {},
};

function getBar(wrapper: RenderResult) {
    return wrapper.container.querySelector(".mx_DecryptionFailureBar");
}

describe("<DecryptionFailureBar />", () => {
    beforeEach(() => {
        jest.useFakeTimers();
    });

    afterEach(() => {
        const container = document.body.firstChild;
        container && document.body.removeChild(container);

        jest.runOnlyPendingTimers();
        jest.useRealTimers();

        mockClient.cancelAndResendEventRoomKeyRequest.mockClear();
    });

    it("Displays a loading spinner", async () => {
        ourDevice = unverifiedDevice1;
        allDevices = [unverifiedDevice1];

        const bar = render(
            // @ts-ignore
            <MatrixClientContext.Provider value={mockClient}>
                <DecryptionFailureBar
                    failures={[
                        // @ts-ignore
                        mockEvent1,
                    ]}
                />
                ,
            </MatrixClientContext.Provider>,
        );

        await waitFor(() => expect(mockClient.isSecretStored).toHaveBeenCalled());

        expect(getBar(bar)).toMatchSnapshot();

        bar.unmount();
    });

    it("Prompts the user to verify if they have other devices", async () => {
        ourDevice = unverifiedDevice1;
        allDevices = [unverifiedDevice1, verifiedDevice1];
        keyBackup = false;

        const bar = render(
            // @ts-ignore
            <MatrixClientContext.Provider value={mockClient}>
                <DecryptionFailureBar
                    failures={[
                        // @ts-ignore
                        mockEvent1,
                    ]}
                />
                ,
            </MatrixClientContext.Provider>,
        );

        await waitFor(() => expect(mockClient.isSecretStored).toHaveBeenCalled());

        act(() => {
            jest.advanceTimersByTime(5000);
        });
        expect(getBar(bar)).toMatchSnapshot();

        bar.unmount();
    });

    it("Prompts the user to verify if they have backups", async () => {
        ourDevice = unverifiedDevice1;
        allDevices = [unverifiedDevice1];
        keyBackup = true;

        const bar = render(
            // @ts-ignore
            <MatrixClientContext.Provider value={mockClient}>
                <DecryptionFailureBar
                    failures={[
                        // @ts-ignore
                        mockEvent1,
                    ]}
                />
                ,
            </MatrixClientContext.Provider>,
        );

        await waitFor(() => expect(mockClient.isSecretStored).toHaveBeenCalled());

        act(() => {
            jest.advanceTimersByTime(5000);
        });
        expect(getBar(bar)).toMatchSnapshot();

        bar.unmount();
    });

    it("Prompts the user to reset if they have no other verified devices and no backups", async () => {
        ourDevice = unverifiedDevice1;
        allDevices = [unverifiedDevice1, unverifiedDevice2];
        keyBackup = false;

        const bar = render(
            // @ts-ignore
            <MatrixClientContext.Provider value={mockClient}>
                <DecryptionFailureBar
                    failures={[
                        // @ts-ignore
                        mockEvent1,
                    ]}
                />
                ,
            </MatrixClientContext.Provider>,
        );

        await waitFor(() => expect(mockClient.isSecretStored).toHaveBeenCalled());

        act(() => {
            jest.advanceTimersByTime(5000);
        });
        expect(getBar(bar)).toMatchSnapshot();

        bar.unmount();
    });

    it("Recommends opening other devices if there are other verified devices", async () => {
        ourDevice = verifiedDevice1;
        allDevices = [verifiedDevice1, verifiedDevice2];
        keyBackup = false;

        const bar = render(
            // @ts-ignore
            <MatrixClientContext.Provider value={mockClient}>
                <DecryptionFailureBar
                    failures={[
                        // @ts-ignore
                        mockEvent1,
                    ]}
                />
                ,
            </MatrixClientContext.Provider>,
        );

        await waitFor(() => expect(mockClient.isSecretStored).toHaveBeenCalled());

        act(() => {
            jest.advanceTimersByTime(5000);
        });
        expect(getBar(bar)).toMatchSnapshot();

        bar.unmount();
    });

    it("Displays a general error message if there are no other verified devices", async () => {
        ourDevice = verifiedDevice1;
        allDevices = [verifiedDevice1, unverifiedDevice1];
        keyBackup = true;

        const bar = render(
            // @ts-ignore
            <MatrixClientContext.Provider value={mockClient}>
                <DecryptionFailureBar
                    failures={[
                        // @ts-ignore
                        mockEvent1,
                    ]}
                />
                ,
            </MatrixClientContext.Provider>,
        );

        await waitFor(() => expect(mockClient.isSecretStored).toHaveBeenCalled());

        act(() => {
            jest.advanceTimersByTime(5000);
        });
        expect(getBar(bar)).toMatchSnapshot();

        bar.unmount();
    });

    it("Displays button to resend key requests if we are verified", async () => {
        ourDevice = verifiedDevice1;
        allDevices = [verifiedDevice1, verifiedDevice2];

        const bar = render(
            // @ts-ignore
            <MatrixClientContext.Provider value={mockClient}>
                <DecryptionFailureBar
                    failures={[
                        // @ts-ignore
                        mockEvent1,
                        // @ts-ignore
                        mockEvent2,
                        // @ts-ignore
                        mockEvent3,
                    ]}
                />
                ,
            </MatrixClientContext.Provider>,
        );

        await waitFor(() => expect(mockClient.isSecretStored).toHaveBeenCalled());

        act(() => {
            jest.advanceTimersByTime(5000);
        });
        expect(getBar(bar)).toMatchSnapshot();

        fireEvent.click(screen.getByText("Resend key requests"));

        expect(mockClient.cancelAndResendEventRoomKeyRequest).toHaveBeenCalledTimes(2);
        expect(mockClient.cancelAndResendEventRoomKeyRequest).toHaveBeenCalledWith(mockEvent1);
        expect(mockClient.cancelAndResendEventRoomKeyRequest).toHaveBeenCalledWith(mockEvent2);

        expect(getBar(bar)).toMatchSnapshot();

        bar.unmount();
    });

    it("Does not display a button to send key requests if we are unverified", async () => {
        ourDevice = unverifiedDevice1;
        allDevices = [unverifiedDevice1, verifiedDevice2];

        const bar = render(
            // @ts-ignore
            <MatrixClientContext.Provider value={mockClient}>
                <DecryptionFailureBar
                    failures={[
                        // @ts-ignore
                        mockEvent1,
                        // @ts-ignore
                        mockEvent2,
                    ]}
                />
                ,
            </MatrixClientContext.Provider>,
        );

        await waitFor(() => expect(mockClient.isSecretStored).toHaveBeenCalled());

        act(() => {
            jest.advanceTimersByTime(5000);
        });
        expect(getBar(bar)).toMatchSnapshot();

        bar.unmount();
    });

    it("Displays the button to resend key requests only if there are sessions we haven't already requested", async () => {
        ourDevice = verifiedDevice1;
        allDevices = [verifiedDevice1, verifiedDevice2];

        const bar = render(
            // @ts-ignore
            <MatrixClientContext.Provider value={mockClient}>
                <DecryptionFailureBar
                    failures={[
                        // @ts-ignore
                        mockEvent3,
                    ]}
                />
                ,
            </MatrixClientContext.Provider>,
        );

        await waitFor(() => expect(mockClient.isSecretStored).toHaveBeenCalled());

        act(() => {
            jest.advanceTimersByTime(5000);
        });
        expect(getBar(bar)).toMatchSnapshot();

        fireEvent.click(screen.getByText("Resend key requests"));

        expect(mockClient.cancelAndResendEventRoomKeyRequest).toHaveBeenCalledTimes(1);
        expect(mockClient.cancelAndResendEventRoomKeyRequest).toHaveBeenCalledWith(mockEvent3);

        expect(getBar(bar)).toMatchSnapshot();

        bar.rerender(
            // @ts-ignore
            <MatrixClientContext.Provider value={mockClient}>
                <DecryptionFailureBar
                    failures={[
                        // @ts-ignore
                        mockEvent1,
                        // @ts-ignore
                        mockEvent2,
                        // @ts-ignore
                        mockEvent3,
                    ]}
                />
                ,
            </MatrixClientContext.Provider>,
        );

        expect(getBar(bar)).toMatchSnapshot();

        mockClient.cancelAndResendEventRoomKeyRequest.mockClear();

        fireEvent.click(screen.getByText("Resend key requests"));

        expect(mockClient.cancelAndResendEventRoomKeyRequest).toHaveBeenCalledTimes(1);
        expect(mockClient.cancelAndResendEventRoomKeyRequest).toHaveBeenCalledWith(mockEvent1);

        expect(getBar(bar)).toMatchSnapshot();

        bar.unmount();
    });

    it("Handles device updates", async () => {
        ourDevice = unverifiedDevice1;
        allDevices = [unverifiedDevice1, verifiedDevice2];

        const bar = render(
            // @ts-ignore
            <MatrixClientContext.Provider value={mockClient}>
                <DecryptionFailureBar
                    failures={[
                        // @ts-ignore
                        mockEvent1,
                        // @ts-ignore
                        mockEvent2,
                    ]}
                />
                ,
            </MatrixClientContext.Provider>,
        );

        await waitFor(() => expect(mockClient.isSecretStored).toHaveBeenCalled());
        act(() => {
            jest.advanceTimersByTime(5000);
        });
        expect(getBar(bar)).toMatchSnapshot();

        ourDevice = verifiedDevice1;
        await act(callback);
        expect(getBar(bar)).toMatchSnapshot();

        bar.unmount();
    });
});
