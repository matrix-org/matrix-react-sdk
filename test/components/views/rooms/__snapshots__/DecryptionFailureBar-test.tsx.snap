// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DecryptionFailureBar /> Displays a general error message if there are no other verified devices 1`] = `
<div
  class="mx_DecryptionFailureBar"
>
  <div
    class="mx_DecryptionFailureBar_icon"
  />
  <div
    class="mx_DecryptionFailureBar_message"
  >
    <div
      class="mx_DecryptionFailureBar_message_headline"
    >
      Some messages could not be decrypted
    </div>
    <div
      class="mx_DecryptionFailureBar_message_body"
    >
      Unfortunately, there are no other verified devices to request decryption keys from. Signing in and verifying other devices may help avoid this situation in the future.
    </div>
  </div>
  <div
    class="mx_DecryptionFailureBar_button"
  />
</div>
`;

exports[`<DecryptionFailureBar /> Displays a loading spinner 1`] = `
<div
  class="mx_DecryptionFailureBar"
>
  <div
    class="mx_Spinner"
  >
    <div
      aria-label="Loading…"
      class="mx_Spinner_icon"
      data-testid="spinner"
      role="progressbar"
      style="width: 32px; height: 32px;"
    />
  </div>
  <div
    class="mx_DecryptionFailureBar_message"
  >
    <div
      class="mx_DecryptionFailureBar_message_headline"
    >
      Decrypting messages…
    </div>
    <div
      class="mx_DecryptionFailureBar_message_body"
    >
      Please wait as we try to decrypt your messages. This may take a few moments.
    </div>
  </div>
  <div
    class="mx_DecryptionFailureBar_button"
  />
</div>
`;

exports[`<DecryptionFailureBar /> Displays button to resend key requests if we are verified 1`] = `
<div
  class="mx_DecryptionFailureBar"
>
  <div
    class="mx_DecryptionFailureBar_icon"
  />
  <div
    class="mx_DecryptionFailureBar_message"
  >
    <div
      class="mx_DecryptionFailureBar_message_headline"
    >
      Open another device to load encrypted messages
    </div>
    <div
      class="mx_DecryptionFailureBar_message_body"
    >
      This device is requesting decryption keys from your other devices. Opening one of your other devices may speed this up.
    </div>
  </div>
  <div
    class="mx_DecryptionFailureBar_button"
  >
    <div
      class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary_outline"
      role="button"
      tabindex="0"
    >
      View your device list
    </div>
  </div>
  <div
    class="mx_DecryptionFailureBar_button"
  >
    <div
      class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary"
      role="button"
      tabindex="0"
    >
      Resend key requests
    </div>
  </div>
</div>
`;

exports[`<DecryptionFailureBar /> Displays button to resend key requests if we are verified 2`] = `
<div
  class="mx_DecryptionFailureBar"
>
  <div
    class="mx_DecryptionFailureBar_icon"
  />
  <div
    class="mx_DecryptionFailureBar_message"
  >
    <div
      class="mx_DecryptionFailureBar_message_headline"
    >
      Open another device to load encrypted messages
    </div>
    <div
      class="mx_DecryptionFailureBar_message_body"
    >
      This device is requesting decryption keys from your other devices. Opening one of your other devices may speed this up.
    </div>
  </div>
  <div
    class="mx_DecryptionFailureBar_button"
  >
    <div
      class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary_outline"
      role="button"
      tabindex="0"
    >
      View your device list
    </div>
  </div>
</div>
`;

exports[`<DecryptionFailureBar /> Displays the button to resend key requests only if there are sessions we haven't already requested 1`] = `
<div
  class="mx_DecryptionFailureBar"
>
  <div
    class="mx_DecryptionFailureBar_icon"
  />
  <div
    class="mx_DecryptionFailureBar_message"
  >
    <div
      class="mx_DecryptionFailureBar_message_headline"
    >
      Open another device to load encrypted messages
    </div>
    <div
      class="mx_DecryptionFailureBar_message_body"
    >
      This device is requesting decryption keys from your other devices. Opening one of your other devices may speed this up.
    </div>
  </div>
  <div
    class="mx_DecryptionFailureBar_button"
  >
    <div
      class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary_outline"
      role="button"
      tabindex="0"
    >
      View your device list
    </div>
  </div>
  <div
    class="mx_DecryptionFailureBar_button"
  >
    <div
      class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary"
      role="button"
      tabindex="0"
    >
      Resend key requests
    </div>
  </div>
</div>
`;

exports[`<DecryptionFailureBar /> Displays the button to resend key requests only if there are sessions we haven't already requested 2`] = `
<div
  class="mx_DecryptionFailureBar"
>
  <div
    class="mx_DecryptionFailureBar_icon"
  />
  <div
    class="mx_DecryptionFailureBar_message"
  >
    <div
      class="mx_DecryptionFailureBar_message_headline"
    >
      Open another device to load encrypted messages
    </div>
    <div
      class="mx_DecryptionFailureBar_message_body"
    >
      This device is requesting decryption keys from your other devices. Opening one of your other devices may speed this up.
    </div>
  </div>
  <div
    class="mx_DecryptionFailureBar_button"
  >
    <div
      class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary_outline"
      role="button"
      tabindex="0"
    >
      View your device list
    </div>
  </div>
</div>
`;

exports[`<DecryptionFailureBar /> Displays the button to resend key requests only if there are sessions we haven't already requested 3`] = `
<div
  class="mx_DecryptionFailureBar"
>
  <div
    class="mx_DecryptionFailureBar_icon"
  />
  <div
    class="mx_DecryptionFailureBar_message"
  >
    <div
      class="mx_DecryptionFailureBar_message_headline"
    >
      Open another device to load encrypted messages
    </div>
    <div
      class="mx_DecryptionFailureBar_message_body"
    >
      This device is requesting decryption keys from your other devices. Opening one of your other devices may speed this up.
    </div>
  </div>
  <div
    class="mx_DecryptionFailureBar_button"
  >
    <div
      class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary_outline"
      role="button"
      tabindex="0"
    >
      View your device list
    </div>
  </div>
  <div
    class="mx_DecryptionFailureBar_button"
  >
    <div
      class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary"
      role="button"
      tabindex="0"
    >
      Resend key requests
    </div>
  </div>
</div>
`;

exports[`<DecryptionFailureBar /> Displays the button to resend key requests only if there are sessions we haven't already requested 4`] = `
<div
  class="mx_DecryptionFailureBar"
>
  <div
    class="mx_DecryptionFailureBar_icon"
  />
  <div
    class="mx_DecryptionFailureBar_message"
  >
    <div
      class="mx_DecryptionFailureBar_message_headline"
    >
      Open another device to load encrypted messages
    </div>
    <div
      class="mx_DecryptionFailureBar_message_body"
    >
      This device is requesting decryption keys from your other devices. Opening one of your other devices may speed this up.
    </div>
  </div>
  <div
    class="mx_DecryptionFailureBar_button"
  >
    <div
      class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary_outline"
      role="button"
      tabindex="0"
    >
      View your device list
    </div>
  </div>
</div>
`;

exports[`<DecryptionFailureBar /> Does not display a button to send key requests if we are unverified 1`] = `
<div
  class="mx_DecryptionFailureBar"
>
  <div
    class="mx_DecryptionFailureBar_icon"
  />
  <div
    class="mx_DecryptionFailureBar_message"
  >
    <div
      class="mx_DecryptionFailureBar_message_headline"
    >
      Verify this device to access all messages
    </div>
    <div
      class="mx_DecryptionFailureBar_message_body"
    >
      This device was unable to decrypt some messages because it has not been verified yet.
    </div>
  </div>
  <div
    class="mx_DecryptionFailureBar_button"
  >
    <div
      class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary"
      role="button"
      tabindex="0"
    >
      Verify
    </div>
  </div>
</div>
`;

exports[`<DecryptionFailureBar /> Handles device updates 1`] = `
<div
  class="mx_DecryptionFailureBar"
>
  <div
    class="mx_DecryptionFailureBar_icon"
  />
  <div
    class="mx_DecryptionFailureBar_message"
  >
    <div
      class="mx_DecryptionFailureBar_message_headline"
    >
      Verify this device to access all messages
    </div>
    <div
      class="mx_DecryptionFailureBar_message_body"
    >
      This device was unable to decrypt some messages because it has not been verified yet.
    </div>
  </div>
  <div
    class="mx_DecryptionFailureBar_button"
  >
    <div
      class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary"
      role="button"
      tabindex="0"
    >
      Verify
    </div>
  </div>
</div>
`;

exports[`<DecryptionFailureBar /> Handles device updates 2`] = `
<div
  class="mx_DecryptionFailureBar"
>
  <div
    class="mx_DecryptionFailureBar_icon"
  />
  <div
    class="mx_DecryptionFailureBar_message"
  >
    <div
      class="mx_DecryptionFailureBar_message_headline"
    >
      Open another device to load encrypted messages
    </div>
    <div
      class="mx_DecryptionFailureBar_message_body"
    >
      This device is requesting decryption keys from your other devices. Opening one of your other devices may speed this up.
    </div>
  </div>
  <div
    class="mx_DecryptionFailureBar_button"
  >
    <div
      class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary_outline"
      role="button"
      tabindex="0"
    >
      View your device list
    </div>
  </div>
  <div
    class="mx_DecryptionFailureBar_button"
  >
    <div
      class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary"
      role="button"
      tabindex="0"
    >
      Resend key requests
    </div>
  </div>
</div>
`;

exports[`<DecryptionFailureBar /> Prompts the user to reset if they have no other verified devices and no backups 1`] = `
<div
  class="mx_DecryptionFailureBar"
>
  <div
    class="mx_DecryptionFailureBar_icon"
  />
  <div
    class="mx_DecryptionFailureBar_message"
  >
    <div
      class="mx_DecryptionFailureBar_message_headline"
    >
      Reset your keys to prevent future decryption errors
    </div>
    <div
      class="mx_DecryptionFailureBar_message_body"
    >
      You will not be able to access old undecryptable messages, but resetting your keys will allow you to receive new messages.
    </div>
  </div>
  <div
    class="mx_DecryptionFailureBar_button"
  >
    <div
      class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary"
      role="button"
      tabindex="0"
    >
      Reset
    </div>
  </div>
</div>
`;

exports[`<DecryptionFailureBar /> Prompts the user to verify if they have backups 1`] = `
<div
  class="mx_DecryptionFailureBar"
>
  <div
    class="mx_DecryptionFailureBar_icon"
  />
  <div
    class="mx_DecryptionFailureBar_message"
  >
    <div
      class="mx_DecryptionFailureBar_message_headline"
    >
      Verify this device to access all messages
    </div>
    <div
      class="mx_DecryptionFailureBar_message_body"
    >
      This device was unable to decrypt some messages because it has not been verified yet.
    </div>
  </div>
  <div
    class="mx_DecryptionFailureBar_button"
  >
    <div
      class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary"
      role="button"
      tabindex="0"
    >
      Verify
    </div>
  </div>
</div>
`;

exports[`<DecryptionFailureBar /> Prompts the user to verify if they have other devices 1`] = `
<div
  class="mx_DecryptionFailureBar"
>
  <div
    class="mx_DecryptionFailureBar_icon"
  />
  <div
    class="mx_DecryptionFailureBar_message"
  >
    <div
      class="mx_DecryptionFailureBar_message_headline"
    >
      Verify this device to access all messages
    </div>
    <div
      class="mx_DecryptionFailureBar_message_body"
    >
      This device was unable to decrypt some messages because it has not been verified yet.
    </div>
  </div>
  <div
    class="mx_DecryptionFailureBar_button"
  >
    <div
      class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary"
      role="button"
      tabindex="0"
    >
      Verify
    </div>
  </div>
</div>
`;

exports[`<DecryptionFailureBar /> Recommends opening other devices if there are other verified devices 1`] = `
<div
  class="mx_DecryptionFailureBar"
>
  <div
    class="mx_DecryptionFailureBar_icon"
  />
  <div
    class="mx_DecryptionFailureBar_message"
  >
    <div
      class="mx_DecryptionFailureBar_message_headline"
    >
      Open another device to load encrypted messages
    </div>
    <div
      class="mx_DecryptionFailureBar_message_body"
    >
      This device is requesting decryption keys from your other devices. Opening one of your other devices may speed this up.
    </div>
  </div>
  <div
    class="mx_DecryptionFailureBar_button"
  >
    <div
      class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary_outline"
      role="button"
      tabindex="0"
    >
      View your device list
    </div>
  </div>
  <div
    class="mx_DecryptionFailureBar_button"
  >
    <div
      class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary"
      role="button"
      tabindex="0"
    >
      Resend key requests
    </div>
  </div>
</div>
`;
